Update click/auxclick/contextmenu and click() to use PointerEvent

rr-codes opened the pull request:
#### 3581a1e578bbe034efda7d6030d7bc528af6db8a
<pre>
Update click/auxclick/contextmenu and click() to use PointerEvent
<a href="https://bugs.webkit.org/show_bug.cgi?id=218665">https://bugs.webkit.org/show_bug.cgi?id=218665</a>
<a href="https://rdar.apple.com/71202646">rdar://71202646</a>

Reviewed by Abrar Rahman Protyasha.

Change the `click`, `auxclick`, and `contextmenu` events to be `PointerEvent`s instead of only
`MouseEvent`s.

Additionally, make the `click()` function dispatch a `PointerEvent` instead of a `MouseEvent`.

* LayoutTests/accessibility/links-activation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/activation-trigger-pointerevent_touch-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_auxclick_is_a_pointerevent_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_click_is_a_pointerevent_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_click_is_a_pointerevent_multiple_clicks_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_click_is_a_pointerevent_multiple_clicks_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_contextmenu_is_a_pointerevent_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/interface/click-event-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/semantics/disabled-elements/disabled-event-dispatch.tentative-expected.txt:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/mac/fast/events/objc-event-api-expected.txt:
* Source/WebCore/dom/Element.cpp:
(WebCore::dispatchPointerEventIfNeeded):
(WebCore::Element::dispatchMouseEvent):
* Source/WebCore/dom/EventDispatcher.cpp:
(WebCore::EventDispatcher::dispatchEvent):
* Source/WebCore/dom/MouseRelatedEvent.cpp:
(WebCore::MouseRelatedEvent::MouseRelatedEvent):
* Source/WebCore/dom/MouseRelatedEvent.h:
(WebCore::MouseRelatedEvent::windowLocation const):
* Source/WebCore/dom/PointerEvent.cpp:
(WebCore::PointerEvent::typeFromMouseEventType):
(WebCore::PointerEvent::create):
(WebCore::PointerEvent::PointerEvent):
(WebCore::pointerEventType): Deleted.
* Source/WebCore/dom/PointerEvent.h:
(WebCore::PointerEvent::buttonForType):
(WebCore::PointerEvent::typeRequiresResolvedButton):
(WebCore::PointerEvent::typeIsUpOrDown): Deleted.
* Source/WebCore/dom/SimulatedClick.cpp:
(WebCore::simulatePointerEvent):
(WebCore::simulateClick):
* Source/WebCore/page/PointerCaptureController.cpp:
(WebCore::PointerCaptureController::pointerEventForMouseEvent):
(WebCore::pointerEventType): Deleted.
* Source/WebKit/Shared/WebMouseEvent.cpp:
(WebKit::mouseButton):

Canonical link: <a href="https://commits.webkit.org/282524@main">https://commits.webkit.org/282524@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2dccc6e953cc0976dd5e685d04931c92068f8a94

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63279 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42635 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15876 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67300 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13887 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65399 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50323 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14167 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50973 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9587 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66348 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39584 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54796 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31656 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36267 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12145 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12759 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57806 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12473 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68996 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7226 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12076 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58285 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7257 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54866 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58515 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6018 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9578 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38456 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39535 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40647 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39278 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/dom/SimulatedClick.cpp:
```diff
@@ -77,11 +79,55 @@ class SimulatedMouseEvent final : public MouseEvent {
+    }
+
+private:
+    // If the device type cannot be detected by the user agent, then the value MUST be an empty string.
```
Are these comments from the spec? If so, might be nice to add a spec link above line 82.

rr-codes replied:
they are; will do

pxlcoder commented on Source/WebCore/dom/SimulatedClick.cpp:
```diff
@@ -77,11 +79,55 @@ class SimulatedMouseEvent final : public MouseEvent {
+    {
+        setUnderlyingEvent(underlyingEvent.get());
+
+        if (auto* pointerEvent = dynamicDowncast<PointerEvent>(this->underlyingEvent())) {
```
```suggestion
        if (RefPtr pointerEvent = dynamicDowncast<PointerEvent>(this->underlyingEvent())) {
```

aprotyas commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -461,13 +461,12 @@ static ShouldIgnoreMouseEvent dispatchPointerEventIfNeeded(Element& element, con
         UNUSED_PARAM(platformEvent);
 #endif
-        if (platformEvent.syntheticClickType() != SyntheticClickType::NoTap)
-            return ShouldIgnoreMouseEvent::No;
```
I'm guessing this is safe to do now that the click event is a PointerEvent? But would we still not face the redundant event issue we tried to address in https://commits.webkit.org/218757@main?

aprotyas replied:
Maybe @whsieh remembers?

whsieh replied:
Right — please double check that a single tap over any element with a `click` event handler doesn't emit redundant pointerdown/pointerup events.

rr-codes replied:
I guess I can just leave this here and just add a check for click/auxclick/contextmenu instead, to be safe

aprotyas commented on Source/WebCore/dom/SimulatedClick.cpp:
```diff
@@ -77,11 +79,55 @@ class SimulatedMouseEvent final : public MouseEvent {
+    static constexpr auto pointerType = ASCIILiteral { ""_s };
+
+    // The pointerId value of -1 MUST be reserved and used to indicate events that were generated by something other than a pointing device.
+    static constexpr auto pointerID = static_cast<PointerID>(-1);
```
We should fix this separately, but this will be wrong. Currently this ID will suffer from integer underflow. The spec says `pointerID` should be a `long` value, i.e. `int32_t`. However, in [263829@main](https://commits.webkit.org/263829@main), I naively changed that to be a `uint32_t` instead to deal with ID overflows. Back then I hadn't realized that negative IDs were valid. We should consider bumping the ID type to be an `int64_t`, instead.

rr-codes replied:
hm, there is explicitly a WPT test that checks if it's -1, and it passes, somehow

aprotyas replied:
Interesting!

aprotyas replied:
Do you also locally log a -1 ID for these events?

aprotyas commented on Source/WebCore/dom/SimulatedClick.cpp:
```diff
@@ -77,11 +79,55 @@ class SimulatedMouseEvent final : public MouseEvent {
+class SimulatedPointerEvent final : public PointerEvent {
+    WTF_MAKE_TZONE_OR_ISO_ALLOCATED_INLINE(SimulatedPointerEvent);
+public:
+    static RefPtr<SimulatedPointerEvent> create(const AtomString& type, const SimulatedMouseEvent& event, RefPtr<Event>&& underlyingEvent, Element& target, SimulatedClickSource source)
```
Could we make this return a `Ref<>` instead?

aprotyas reviewed (approved):
Looks good to me! One question about the synthetic click type, and one to-do we should address separately.

nt1m commented on LayoutTests/platform/mac/TestExpectations:
```diff
@@ -1414,13 +1414,13 @@ imported/w3c/web-platform-tests/pointerevents/compat/pointerevent_mouseevent_key
 imported/w3c/web-platform-tests/pointerevents/pointerevent_capture_suppressing_mouse.html [ Skip ]
 imported/w3c/web-platform-tests/pointerevents/pointerevent_change-touch-action-onpointerdown_touch.html [ Skip ]
-imported/w3c/web-platform-tests/pointerevents/pointerevent_click_is_a_pointerevent.html?mouse [ Skip ]
+imported/w3c/web-platform-tests/pointerevents/pointerevent_click_is_a_pointerevent.html?mouse [ Pass ]
```
Does it need to be pass or can it just be removed?

aprotyas replied:
Yes I think it can just be removed https://github.com/WebKit/WebKit/blob/ba3e2658de9289485203a58829140f7ae0193a48/LayoutTests/platform/mac/TestExpectations#L93

rr-codes committed:
Update click/auxclick/contextmenu and click() to use PointerEvent
https://bugs.webkit.org/show_bug.cgi?id=218665
rdar://71202646

Reviewed by Abrar Rahman Protyasha.

Change the `click`, `auxclick`, and `contextmenu` events to be `PointerEvent`s instead of only
`MouseEvent`s.

Additionally, make the `click()` function dispatch a `PointerEvent` instead of a `MouseEvent`.

* LayoutTests/accessibility/links-activation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/activation-trigger-pointerevent_touch-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_auxclick_is_a_pointerevent_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_click_is_a_pointerevent_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_click_is_a_pointerevent_multiple_clicks_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_click_is_a_pointerevent_multiple_clicks_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_contextmenu_is_a_pointerevent_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/interface/click-event-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/semantics/disabled-elements/disabled-event-dispatch.tentative-expected.txt:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/mac/fast/events/objc-event-api-expected.txt:
* Source/WebCore/dom/Element.cpp:
(WebCore::dispatchPointerEventIfNeeded):
(WebCore::Element::dispatchMouseEvent):
* Source/WebCore/dom/EventDispatcher.cpp:
(WebCore::EventDispatcher::dispatchEvent):
* Source/WebCore/dom/MouseRelatedEvent.cpp:
(WebCore::MouseRelatedEvent::MouseRelatedEvent):
* Source/WebCore/dom/MouseRelatedEvent.h:
(WebCore::MouseRelatedEvent::windowLocation const):
* Source/WebCore/dom/PointerEvent.cpp:
(WebCore::PointerEvent::typeFromMouseEventType):
(WebCore::PointerEvent::create):
(WebCore::PointerEvent::PointerEvent):
(WebCore::pointerEventType): Deleted.
* Source/WebCore/dom/PointerEvent.h:
(WebCore::PointerEvent::buttonForType):
(WebCore::PointerEvent::typeRequiresResolvedButton):
(WebCore::PointerEvent::typeIsUpOrDown): Deleted.
* Source/WebCore/dom/SimulatedClick.cpp:
(WebCore::simulatePointerEvent):
(WebCore::simulateClick):
* Source/WebCore/page/PointerCaptureController.cpp:
(WebCore::PointerCaptureController::pointerEventForMouseEvent):
(WebCore::pointerEventType): Deleted.
* Source/WebKit/Shared/WebMouseEvent.cpp:
(WebKit::mouseButton):

Canonical link: https://commits.webkit.org/282524@main
